core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizerUnitTests.kt

```diff
@@ -1,190 +1,882 @@
+
+        // do a sync pass, addressing the conflict
+        ctx.doSyncPass()
+        ctx.waitForEvent()
+        // verify that a change event has been emitted. the conflict will have been handled
+        // in setupPendingReplace
+        ctx.verifyChangeEventListenerCalledForActiveDoc(
+            times = 1,
+            expectedChangeEvent = ChangeEvent.changeEventForLocalInsert(
+                ctx.namespace, expectedDoc, false
+            ))
+        ctx.verifyConflictHandlerCalledForActiveDoc(times = 0)
+        ctx.verifyErrorListenerCalledForActiveDoc(times = 0)
+
+        assertEquals(expectedDoc, ctx.findTestDocumentFromLocalCollection())
+
+        // 2: Replace -> Conflict -> Delete (remote wins)
+        ctx = harness.freshTestContext()
+        expectedDoc = BsonDocument("count", BsonInt32(3)).append("_id", ctx.testDocumentId)
+        setupPendingReplace(ctx, expectedDoc, shouldConflictBeResolvedByRemote = true)
+
+        ctx.verifyConflictHandlerCalledForActiveDoc(times = 1)
+        ctx.verifyErrorListenerCalledForActiveDoc(times = 0)
+        assertNull(ctx.findTestDocumentFromLocalCollection())
+
+        // 3: Replace -> Conflict -> Exception -> Freeze
+        ctx = harness.freshTestContext()
+        expectedDoc = BsonDocument("count", BsonInt32(3)).append("_id", ctx.testDocumentId)
+        ctx.exceptionToThrowDuringConflict = Exception("bad")
+        // verify that, though the conflict handler was called, the exceptionToThrow was emitted
+        // by the dataSynchronizer
+        setupPendingReplace(ctx, expectedDoc, shouldWaitForError = true)
+        assertEquals(expectedDoc, ctx.findTestDocumentFromLocalCollection())
+
+        // clear issues. open a path for a delete.
+        // do another sync pass. the doc should remain the same as it is frozen
+        ctx.exceptionToThrowDuringConflict = null
+        ctx.shouldConflictBeResolvedByRemote = false
+        ctx.doSyncPass()
+        assertEquals(expectedDoc, ctx.findTestDocumentFromLocalCollection())
+        expectedDoc = BsonDocument("count", BsonInt32(5)).append("_id", ctx.testDocumentId)
+
+        // replace the doc locally (with an update), unfreezing it, and syncing it
+        setupPendingReplace(ctx, expectedDoc)
+        ctx.doSyncPass()
+        assertEquals(expectedDoc, ctx.findTestDocumentFromLocalCollection())
+
+        // 4: Unknown -> Freeze
+        ctx = harness.freshTestContext()
+        expectedDoc = BsonDocument("count", BsonInt32(3)).append("_id", ctx.testDocumentId)
+        ctx.queueConsumableRemoteUnknownEvent()
+        setupPendingReplace(ctx, expectedDoc)
+
+        ctx.queueConsumableRemoteUpdateEvent()
+        ctx.doSyncPass()
+        assertEquals(expectedDoc, ctx.findTestDocumentFromLocalCollection())
+
+        // should be frozen since the operation type was unknown
+        ctx.queueConsumableRemoteUnknownEvent()
+        ctx.doSyncPass()
+        assertEquals(expectedDoc, ctx.findTestDocumentFromLocalCollection())
```

Comment:
[nit q] how are we actually testing that the document is frozen? shouldn't we try remote deleting it first before checking that it's still there?

Reply:
I can add a case for this, though I believe I handle it with the previous update. There is also no "real" remote– we're testing this by queueing fake events as if they came from the remote.